README.md
#PHP Yandex.Money API SDK

## Prerequisites

* PHP 5.3 or above
* curl, json & openssl extensions must be enabled
* composer for fetching dependencies (See http://getcomposer.org)

## Links

* Yandex.Money API page: [Ru](http://api.yandex.ru/money/), [En](http://api.yandex.com/money/)
* [sample app](https://github.com/yandex-money/yandex-money-sdk-php/tree/master/sample)
* [changelog](https://github.com/yandex-money/yandex-money-sdk-php/blob/master/CHANGELOG.md)

## Getting started

### Workflow

Your app asks user to give permissions to manage operations with user's Yandex.Money account. This process is known as
OAuth authorization. As a result of OAuth your app receives access token from Yandex.Money servers.  

Then, with this token your app can make requests to our servers and perform operations with user's account, sometimes even without
his actual activity.

### Code samples

#### OAuth and token receiving

First of all, we should get token. It's very simple. We take special uri which we'll sent user to ask permissions.

```php
$scope = "account-info operation-history operation-details "; // this is scope of permissions
$authUri = YandexMoney::authorizeUri(YOUR_APP_CLIENT_ID, YOUR_APP_REDIRECT_URI, $scope);
header('Location: ' . $authUri);
```

To get more information about permissions scope please visit Yandex.Money [API docs](http://api.yandex.com/money/doc/dg/concepts/protocol-rights.xml).

`YOUR_APP_CLIENT_ID` and `YOUR_APP_REDIRECT_URI` are parameters that you get when [register](https://sp-money.yandex.ru/myservices/new.xml) your app in Yandex.Money API.

Then Yandex.Money redirect user back to your app. At this redirect uri we should change temporary code parameter from
GET redirect request. This is even more simple.

```php
$ym = new YandexMoney(YOUR_APP_CLIENT_ID);
$receiveTokenResp = $ym->receiveOAuthToken($code, YOUR_APP_REDIRECT_URI, YOUR_APP_CLIENT_SECRET);

if ($receiveTokenResp->isSuccess()) {
    $token = $receiveTokenResp->getAccessToken();
    ... // Here you can store received token to your app's storage
} else {
    print "Error: " . $receiveTokenResp->getError();
    ...
}
```

#### Account info request

```php
$resp = $ym->accountInfo($token);
if ($resp->isSuccess()) {
    print $resp->getBalance();
    print $resp->getAccount();
    ...
} else {
    print "Error: " . $resp->getError();
    die();
}
```

Rest of requests you can use the same way. After this example we won't show you check on request success.

#### Operation history and details

```php
$resp = $ym->operationHistory($token, 0, 5); 
// second param is first record record from set, third param is record count

$resp = $ym->operationDetail($token, $requestId); 
// second param is requestId from payment method or one from operation hisory
```

#### p2p transfer

```php
$resp = $ym->requestPaymentP2P($token, "410011161616877", "0.02", 
        "comment to sender", "message to recepient");

$requestId = $resp->getRequestId();
// payment by user's Yandex.Money wallet
$resp = $ym->processPaymentByWallet($token, $requestId); 
```

#### Payment to shop by credit card

We'll show it on example of payment to Megafon.

```php
$params["pattern_id"] = "337"; // pattern_id - is id of shop in Yandex.Money.
$params["PROPERTY1"] = "921"; // preffix number
$params["PROPERTY2"] = "3020052"; // phone number
$params["sum"] = "2.00"; // amount
$resp = $ym->requestPaymentShop($token, $params);

$requestId = $resp->getRequestId();
$resp = $ym->processPaymentByCard($token, $requestId, "375"); // third param is cvc of user's credit card
```

#### Logs

We recommend you to log request to Yandex.Money system. It's easy, you should only set log file name and path relatively to your current script.

Notice that if you want to log details of responses you should do it yourself.

```php
$ym = new YandexMoney(YOUR_APP_CLIENT_ID, './path/to/logfile/ym.log');
```


CHANGELOG.md
# Changelog

## 1.3.0 (18.03.2014)

* composer support
* php >= 5.3.0 requirement
* details flag added in operation-history method
* several minor docs and code changes
* sample project improved

## 1.2.5 (27.12.2013)

* certificate chain updated

## 1.2.4 (28.10.2013)

* fixed undefined variable $resp at ApiRequestor

## 1.2.3 (15.08.2013)

* create log file instead of exception throwing

## 1.2.2 (03.08.2013)

* revokeOAuthToken exception fixed

## 1.2.1 (09.01.2013)

* request logs added

## 1.2.0 (29.05.2012)

* revoke token method added

## 1.1.0 (09.04.2012)

* source code refactoring
* credit card payments to shops

## 1.0.0 (28.10.2012)

* php-lib release

sample/index.php
<?php

ini_set('display_errors', 1);

// Include the composer autoloader
if(!file_exists(__DIR__ .'/vendor/autoload.php')) {
    echo "folder: " . __DIR__ . "\r\n";
    echo "The 'vendor' folder is missing. You must run 'composer update' to resolve application dependencies.\nPlease see the README for more information.\n";
    exit(1);
}

require_once __DIR__ . '/vendor/autoload.php';
require_once __DIR__ . '/consts.php';

use YandexMoney\Client;

$code = $_GET['code'];
if (!isset($code)) { // If we are just begginig OAuth
    $scope = "account-info " .
        "operation-history " .
        "operation-details " .
        "payment.to-account(\"410011161616877\",\"account\").limit(30,10) " .
        "payment.to-pattern(\"337\").limit(30,10) " .
        "money-source(\"wallet\",\"card\") ";
    $authUri = Client::authorizeUri(CLIENT_ID, REDIRECT_URI, $scope);
    header('Location: ' . $authUri);

} else { // when we recieved a temporary code on redirect
    ?>

<html>
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <link rel="stylesheet" type="text/css" href="styles.css">
    <title>Yandex.Money PHP SDK sample app</title>
</head>
<body>
<div id="main">
    <h3 id="header">Yandex.Money PHP SDK sample app</h3>

    <?php
        $ym = new Client(CLIENT_ID, __DIR__ . '/ym.log');
        $receiveTokenResp = $ym->receiveOAuthToken($code, REDIRECT_URI, CLIENT_SECRET);

        print "<p class=\"output\">";
        if ($receiveTokenResp->isSuccess()) {
            $token = $receiveTokenResp->getAccessToken();
            print "Received token: " . $token;
        } else {
            print "Error: " . $receiveTokenResp->getError();
            die();
        }
        print "</p>";
        print "<p>Notice: after you received access_token you should store it to your app's storage</p>";


        $resp = $ym->accountInfo($token);
        print "<p class=\"output\">";
        if ($resp->isSuccess()) {
            var_dump($resp);
        } else {
            print "Error: " . $resp->getError();
            die();
        }
        print "</p>";


        $resp = $ym->operationHistory($token, 0, 3);
        print "<p class=\"output\">";
        if ($resp->isSuccess()) {
            var_dump($resp);
        } else {
            print "Error: " . $resp->getError();
            die();
        }
        print "</p>";

        $operations = $resp->getOperations();
        if (count($operations) > 0) {
            $requestId = $operations[0]->getOperationId();
            $resp = $ym->operationDetail($token, $requestId);
            print "<p class=\"output\">";
            if ($resp->isSuccess()) {
                var_dump($resp);
            } else {
                print "Error: " . $resp->getError();
                die();
            }
            print "</p>";    
        } else {
            print "<p>You have no any history records at your account to request details.</p>";
        }
        

        $resp = $ym->requestPaymentP2P($token, "410011161616877", "0.02");
        print "<p class=\"output\">";
        if ($resp->isSuccess()) {
            var_dump($resp);
        } else {
            print "Error: " . $resp->getError();
            die();
        }
        print "</p>";

        $requestId = $resp->getRequestId();
        $resp = $ym->processPaymentByWallet($token, $requestId);
        print "<p class=\"output\">";
        if ($resp->isSuccess()) {
            var_dump($resp);
        } else {
            print "Error: " . $resp->getError();
            die();
        }
        print "</p>";

        $params["pattern_id"] = "337";
        $params["PROPERTY1"] = "921";
        $params["PROPERTY2"] = "3020052";
        $params["sum"] = "2.00";
        $resp = $ym->requestPaymentShop($token, $params);
        print "<p class=\"output\">";
        if ($resp->isSuccess()) {
            var_dump($resp);
        } else {
            print "Error: " . $resp->getError();
            die();
        }
        print "</p>";

        $requestId = $resp->getRequestId();
        $resp = $ym->processPaymentByCard($token, $requestId, "313");
        print "<p class=\"output\">";
        if ($resp->isSuccess()) {
            var_dump($resp);
        } else {
            print "Error: " . $resp->getError();
            die();
        }
        print "</p>";

        if ($ym->revokeOAuthToken($token)) {
            print "<p>Token was successfully revoked</p>";
        }
    }

    ?>
</div>
</body>
</html>